core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -152,6 +152,9 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
        */
       updatedBlocks ++=
         blockManager.putIterator(key, values, level, tellMaster = true, effectiveStorageLevel)
+      if (level.useMemory) {
+        blockManager.memoryStore.releasePendingUnrollMemoryForThisThread()
+      }
```

Comment:
I think this is correct, but it's a confusing place to do this. It requires the reader of the code to understand the difference between `level` and `putLevel`. I would move this check to after the `putInBlockManager` in L185 of the old code instead.